### PR TITLE
Improve `dq.wigner()` (remove fft method, support arbitrary `xvec` and `yvec`)

### PR DESCRIPTION
**Pull request by abocquet**
Closes https://github.com/dynamiqs/dynamiqs/issues/528.

Closes DYN-150.

I chose to set `xvec` and `yvec` as static arguments because I assume they will not change much and this avoids one conditional branching

### CODE REVIEW

**Review by gautierronan — CHANGES_REQUESTED** [review]
LGTM. I guess we should propagate these changes to related `plot_...` functions as well.

**Comment by pierreguilmin** [comment]
Great, I'll review this evening or tmo!

> LGTM. I guess we should propagate these changes to related plot_... functions as well.
I don't think it makes sense, `xvec` and `yvec` can be arbitrary here (not necessary evenly-spaced, nor ordered).

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
We need to
- add `xvec`, `yvec` to the documentation
- fix the typing: `ArrayLike | None`
- check their shape if they're not `None`

@gautierronan are both methods (clenshaw and fft) working whatever `xvec` and `yvec`? (not evenly spaced, not ordered, repeated elements, ...) If not we should document in a warning what's possible for `xvec`/`yvec`.

**Comment by gautierronan** [comment]
> @gautierronan are both methods (clenshaw and fft) working whatever xvec and yvec? (not evenly spaced, not ordered, repeated elements, ...) If not we should document what's suitable for xvec/yvec.

I think we should remove the `fft` method altogether. It makes sense in qutip because it's much faster, but here clenshaw is jitted so it's really comparable in runtime, and much more flexible. I don't see any reason to use `fft`, and it's annoying to maintain.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
A few more things to fix:
- Remove unused `_fock_to_position`
- Check shape for xvec and yvec
- Fix return documentation
- Fix CI (merge with main, correct line length)

A few extras:
- @gautierronan we discussed this a while ago, I think it was related to the fft method, can we set the default xmax/ymax to e.g. 6.0 instead of this weird factor?
- Should we take in a single array of complex displacements rather than `xvec`, `yvec`? This is even more general, because currently we force to compute on a grid (that can be irregular). @gautierronan would the clenshaw algorithm work in the case of arbitrary displacements?
- Ideally we should support arbitrary batching on `xvec`/`yvec` or on the displacement array (we can do this in another PR).

**Comment by pierreguilmin** [comment]
Also (for another PR), the JIT is probably disallowing the function to work correctly with QuTiP QObj, we should either fix this generally https://linear.app/dynamiqs/issue/DYN-167/allow-jitted-function-to-take-qutip-qobj-as-parameters or replay the trick we used for mesolve/sesolve (separate arrays conversion from jitted backend).

File: dynamiqs/plots/plots_wigner.py [hunk 1]
```diff
@@ -354,3 +359,15 @@ def plot_wigner_gif(
...
             shutil.rmtree(tmpdir, ignore_errors=True)
+
+
+def check_state_device_type(state: ArrayLike) -> ArrayLike:
+    if next(iter(state.devices())).platform == 'gpu' and state.dtype == jnp.complex128:
+        logging.warning(
+            'Wigner function is not supported yet for f64 on GPU. '
+            'Dynamiqs will copy the state to the CPU to compute the wigner function. '
+            'Performance penalty is expected. If this is a problem for you, open an '
+            'issue at https://github.com/dynamiqs/dynamiqs/issues'
```
Comment (gautierronan) [inline dynamiqs/plots/plots_wigner.py:370]:
```suggestion
            'Wigner-related functions are not yet supported for double precision (f64) on GPU. '
            'The `state` array will be copied to CPU to compute the wigner function. '
            'Performance penalty is expected. If this is a problem for you, open an '
            'issue at https://github.com/dynamiqs/dynamiqs/issues'
```

**Review by gautierronan — APPROVED** [review]
LGTM.


> @gautierronan we discussed this a while ago, I think it was related to the fft method, can we set the default xmax/ymax to e.g. 6.0 instead of this weird factor?

Yes we can if we remove `fft` indeed! BTW, this wierd factor is just 2pi :D 

> Should we take in a single array of complex displacements rather than xvec, yvec? This is even more general, because currently we force to compute on a grid (that can be irregular). @gautierronan would the clenshaw algorithm work in the case of arbitrary displacements?

In principle it should work, but it requires a bit of refactoring and rethinking. I'd be in favor of keeping it like that for now, until someone actually asks for it. Especially since people can always compute the expectation value of the displaced parity if needed.

> Ideally we should support arbitrary batching on xvec/yvec or on the displacement array (we can do this in another PR).

Again, a bit niche I think, but shouldn't be too hard to do.

**Comment by pierreguilmin** [comment]
I have a few modifications to propose, but I won't have time before this weekend sorry.

**Comment by abocquet** [comment]
Unless something is really critical let's merge and change later. Or tell me what bothers you and I'll change

**Comment by pierreguilmin** [comment]
Main thing is why did you move the `states = check_state_device_type(states)` check to plot functions?

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**